### PR TITLE
Opt in for beta updates

### LINKED ISSUE #1458: Beta opt in
Option for the user to opt in for beta pre releases released on github.com

### PR DESCRIPTION
**Pull request by 1-alex98**
Fixes #1458

### CODE REVIEW

File: src/main/java/com/faforever/client/config/ClientProperties.java [hunk 1]
```diff
@@ -41,6 +42,14 @@
...
   }
 
+  @Data
+  public static class GitHubRepo {
+    /**
+     * Api URL to the client GitHub Repo
+     */
+    private String apiUrl;
+  }
+
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/config/ClientProperties.java]: I would add this to the remote preferences / client configuration to also taking into account  a potential renaming of the repository (#1288)

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java [hunk 2]
```diff
@@ -134,14 +135,15 @@
   public CheckBox disallowJoinsCheckBox;
   public JFXToggleButton secondaryVaultLocationToggleButton;
   public Button autoJoinChannelsButton;
+  public JFXToggleButton prereleaseToggleButton;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/preferences/ui/SettingsController.java]: I would use ToggleButton instead. Always use the most generic one. (Also applies to line 136)

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java [hunk 3]
```diff
@@ -292,6 +295,13 @@ public void initialize() {
       preferences.getForgedAlliance().setVaultBaseDirectory(vaultBaseDirectory);
     });
 
+    prereleaseToggleButton.selectedProperty().bindBidirectional(preferences.prereleaseCheckEnabledProperty());
+    prereleaseToggleButton.selectedProperty().addListener((observable, oldValue, newValue) -> {
+      if (!oldValue && newValue) {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/preferences/ui/SettingsController.java]: Is this safe for NullPointerExceptions?

File: src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    headers.add("Accept", "application/vnd.github.v3+json");
+    HttpEntity<String> entity = new HttpEntity<>(null, headers);
+
+    ResponseEntity<List<GitHubRelease>> response = restTemplate.exchange(clientProperties.getGitHubRepo().getApiUrl() + PATH_FOR_RELEASE, HttpMethod.GET, entity, new ParameterizedTypeReference<List<GitHubRelease>>() {
+    });
+    List<GitHubRelease> responseBody = response.getBody();
+    if (responseBody == null || responseBody.isEmpty()) {
+      return null;
+    }
+    GitHubRelease latestRelease = responseBody.get(0);
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java]: 1. Shouldn't the RestTemplate come via DI?
2. If we need a new one, create it in the constructor
3. The whole code of setting up the restTemplate call should be extracted into a separate method for better readability.

File: src/main/java/com/faforever/client/update/ClientUpdateService.java [hunk 5]
```diff
@@ -8,13 +8,18 @@
   /**
    * Returns information about an available mandatory update. Returns {@code null} if no update is available.
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/update/ClientUpdateService.java]: I guess this should be renamed as well?

File: src/main/java/com/faforever/client/update/UpdateInfo.java [hunk 6]
```diff
@@ -12,5 +12,6 @@
   private URL url;
   private int size;
   private URL releaseNotesUrl;
+  private boolean prerelease;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/update/UpdateInfo.java:15]: preRelease?

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/config/ClientProperties.java (hunk 1 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/config/ClientProperties.java]: ok

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java (hunk 2 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/preferences/ui/SettingsController.java]: Well i autogenerate those with Intelij. Is it really that bad? I would like to keep using this feature. I mean I could have also used Checkable then but not sure why i should to that in this very case.

File: src/main/java/com/faforever/client/update/UpdateInfo.java (hunk 6 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/update/UpdateInfo.java:15]: it is one word!

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java (hunk 2 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/client/preferences/ui/SettingsController.java]: I would do it, as it makes easier to drop JFoenix if you want to (e.g. different themes).

**Comment by 1-alex98** [comment]
fitting config can be fund here https://content.test.faforever.com/dfc-config.json

File: src/main/java/com/faforever/client/main/MainController.java [hunk 7]
```diff
@@ -476,7 +476,11 @@ private void enterLoggedInState() {
     getMainScene().setContent(mainRoot);
     getMainScene().setMoveControl(mainRoot);
 
-    clientUpdateService.checkForRegularUpdateInBackground();
+    if (preferencesService.getPreferences().isPrereleaseCheckEnabled()) {
+      clientUpdateService.checkForRegularUpdateInBackground();
+    } else {
+      clientUpdateService.checkForBetaUpdateInBackground();
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/main/MainController.java]: Isn't one method with a boolean parameter a better choice? Why create 2 methods?
And why is the preference read in a UI class, shouldn't the clientUpdateService just read the property itself?

File: src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    );
+  }
+
+  private GitHubAssets getAssetOfFileWithEnding(GitHubRelease latestRelease, String ending) {
+    return latestRelease.getAssets().stream()
+        .filter(asset -> asset.getName().contains(ending))
+        .findFirst()
+        .orElseThrow();
+  }
+}
```
Comment (Geosearchef) [inline src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java:82]: Isn't this the same code as in the regular Update task?

**Review by Geosearchef — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/main/MainController.java (hunk 7 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/main/MainController.java]: Well this is after Logging in. But yeah we could have an event here...

File: src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java (hunk 8 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/update/CheckForBetaUpdateTask.java:82]: no the regular update task does not use GitHub at all!

**Review by Brutus5000 — DISMISSED** [review]

File: src/main/java/com/faforever/client/config/ClientProperties.java [hunk 9]
```diff
@@ -152,5 +152,6 @@
     private String applicationId;
     private String smallImageKey;
     private String bigImageKey;
+    private String discordPrereleaseFeedback;
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/config/ClientProperties.java]: Is this a channel name? A URL? Please name variable properly

File: src/main/java/com/faforever/client/preferences/ui/SettingsController.java [hunk 10]
```diff
@@ -508,5 +519,9 @@ public void onUseNoBackgroundImage(ActionEvent actionEvent) {
     preferencesService.getPreferences().getMainWindow().setBackgroundImagePath(null);
     preferencesService.storeInBackground();
   }
+
+  public void openDiscordFeedback() {
```
Comment (Brutus5000) [inline src/main/java/com/faforever/client/preferences/ui/SettingsController.java]: openDiscordFeedbackChannel?

**Review by Brutus5000 — CHANGES_REQUESTED** [review]

**Pull request merged**